results/cmd/watcher/main.go

```diff
@@ -100,20 +108,76 @@ func (r *reconciler) Reconcile(ctx context.Context, key string) error {
 		return err
 	}
 
-	r.logger.Infof("Sending update for %s/%s (uid %s)", namespace, tr.Name, tr.UID)
+	r.logger.Infof("Recieving new TaskRun %s/%s", namespace, tr.Name)
 
 	// Send the new status of the TaskRun to the API server.
 	p, err := convert.ToProto(tr)
 	if err != nil {
 		r.logger.Errorf("Error converting to proto: %v", err)
 		return err
 	}
-	if _, err := r.client.CreateTaskRunResult(ctx, &pb.CreateTaskRunRequest{
-		TaskRun: p,
-	}); err != nil {
-		r.logger.Error("Error updating TaskRun %s: %v", name, err)
-		return err
-	}
 
+	// Create a TaskRun if it does not exist in results server, update existing one otherwise.
+	annotation := p.GetMetadata().GetAnnotations()
+	if val, ok := annotation["results.tekton.dev/id"]; ok {
+		if r.fakeserver == nil {
+			if _, err := r.client.UpdateTaskRunResult(ctx, &pb.UpdateTaskRunRequest{
+				TaskRun:   p,
+				ResultsId: val,
+			}); err != nil {
+				r.logger.Error("Error updating TaskRun %s: %v", name, err)
+				return err
+			}
+		} else {
+			if _, err := r.fakeserver.UpdateTaskRunResult(ctx, &pb.UpdateTaskRunRequest{
+				TaskRun:   p,
+				ResultsId: val,
+			}); err != nil {
+				r.logger.Error("Error updating TaskRun %s: %v", name, err)
+				return err
+			}
+		}
+		r.logger.Infof("Sending updates for TaskRun %s/%s(results_id: %s)", namespace, tr.Name, val)
+	} else {
+		var res *pb.TaskRunResult
+		if r.fakeserver == nil {
+			res, err = r.client.CreateTaskRunResult(ctx, &pb.CreateTaskRunRequest{
+				TaskRun: p,
+			})
+			if err != nil {
+				r.logger.Error("Error creating TaskRun %s: %v", name, err)
+				return err
+			}
+		} else {
+			res, err = r.fakeserver.CreateTaskRunResult(ctx, &pb.CreateTaskRunRequest{
+				TaskRun: p,
+			})
+			if err != nil {
+				r.logger.Error("Error creating TaskRun %s: %v", name, err)
+				return err
+			}
+		}
+		path, err := annotationPath(res.GetResultsId())
+		if err != nil {
+			r.logger.Error("Error jsonpatch for TaskRun %s : %v", name, err)
+			return err
+		}
+		r.pipelineclientset.TektonV1beta1().TaskRuns(namespace).Patch(name, types.JSONPatchType, path)
+		r.logger.Infof("Creating a new TaskRun %s/%s(results_id: %s)", namespace, tr.Name, res.GetResultsId())
+	}
 	return nil
 }
+
+// AnnotationPath creates a jsonpatch path used for adding results_id to TaskRun annotations field.
+func annotationPath(val string) ([]byte, error) {
+	patches := []jsonpatch.JsonPatchOperation{{
+		Operation: "add",
+		Path:      "/metadata/annotations/results.tekton.dev~1id",
```

Comment:
What does this `~1` syntax mean? Is there a way to dervive it from a common const to share it with the string above?

Reply:
It's just an escape character representing /